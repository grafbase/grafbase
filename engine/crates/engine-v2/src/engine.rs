use ::runtime::{
    auth::AccessToken,
    hooks::Hooks,
    hot_cache::{CachedDataKind, HotCache, HotCacheFactory},
};
use async_runtime::stream::StreamExt as _;
use engine::{BatchRequest, Request};
use engine_parser::types::OperationType;
use futures::{channel::mpsc, FutureExt, StreamExt};
use futures_util::{SinkExt, Stream};
use gateway_core::StreamingFormat;
use gateway_v2_auth::AuthService;
use grafbase_tracing::{
    gql_response_status::GraphqlResponseStatus,
    grafbase_client::Client,
    metrics::{GraphqlOperationMetrics, GraphqlOperationMetricsAttributes},
    span::{gql::GqlRequestSpan, GqlRecorderSpanExt, GqlRequestAttributes},
};
use headers::HeaderMapExt;
use schema::Schema;
use std::{borrow::Cow, sync::Arc};
use tracing::Instrument;
use trusted_documents::PreparedOperationDocument;
use web_time::Instant;

use crate::{
    execution::{ExecutableOperation, PreExecutionContext},
    http_response::{HttpGraphqlResponse, HttpGraphqlResponseExtraMetadata},
    operation::{Operation, OperationMetadata, PreparedOperation, Variables},
    response::{ErrorCode, GraphqlError, Response},
    websocket,
};

mod cache;
mod rate_limiting;
mod runtime;
mod trusted_documents;

pub use crate::engine::rate_limiting::RateLimitContext;
pub use runtime::Runtime;

pub(crate) struct SchemaVersion(Vec<u8>);

impl std::ops::Deref for SchemaVersion {
    type Target = [u8];

    fn deref(&self) -> &Self::Target {
        self.0.as_ref()
    }
}

pub struct Engine<R: Runtime> {
    // We use an Arc for the schema to have a self-contained response which may still
    // needs access to the schema strings
    pub(crate) schema: Arc<Schema>,
    pub(crate) schema_version: SchemaVersion,
    pub(crate) runtime: R,
    operation_metrics: GraphqlOperationMetrics,
    auth: AuthService,
    trusted_documents_cache: <R::CacheFactory as HotCacheFactory>::Cache<String>,
    operation_cache: <R::CacheFactory as HotCacheFactory>::Cache<Arc<PreparedOperation>>,
}

impl<R: Runtime> Engine<R> {
    /// schema_version is used in operation cache key which ensures we only retrieve cached
    /// operation for the same schema version. If none is provided, a random one is generated.
    pub async fn new(schema: Arc<Schema>, schema_version: Option<&[u8]>, runtime: R) -> Self {
        let auth = gateway_v2_auth::AuthService::new_v2(
            schema.settings.auth_config.clone().unwrap_or_default(),
            runtime.kv().clone(),
        );

        Self {
            schema,
            schema_version: SchemaVersion({
                let mut out = Vec::new();
                match schema_version {
                    Some(version) => {
                        out.push(0x00);
                        out.extend_from_slice(version);
                    }
                    None => {
                        out.push(0x01);
                        out.extend_from_slice(&ulid::Ulid::new().to_bytes());
                    }
                }
                out
            }),
            auth,
            operation_metrics: GraphqlOperationMetrics::build(runtime.meter()),
            trusted_documents_cache: runtime.cache_factory().create(CachedDataKind::PersistedQuery).await,
            operation_cache: runtime.cache_factory().create(CachedDataKind::Operation).await,
            runtime,
        }
    }

    pub async fn execute(
        self: &Arc<Self>,
        headers: http::HeaderMap,
        batch_request: BatchRequest,
    ) -> HttpGraphqlResponse {
        let format = headers.typed_get::<StreamingFormat>();
        let request_context = match self.create_request_context(headers).await {
            Ok(context) => context,
            Err(response) => return HttpGraphqlResponse::build(response, format, Default::default()),
        };

        if let Err(err) = self.runtime.rate_limiter().limit(&RateLimitContext::Global).await {
            return HttpGraphqlResponse::build(
                Response::pre_execution_error(GraphqlError::new(err.to_string(), ErrorCode::RateLimitError)),
                format,
                Default::default(),
            );
        }

        self.execute_with_context(request_context, batch_request).await
    }

    pub async fn create_session(self: &Arc<Self>, headers: http::HeaderMap) -> Result<Session<R>, Cow<'static, str>> {
        if let Err(err) = self.runtime.rate_limiter().limit(&RateLimitContext::Global).await {
            return Err(
                Response::pre_execution_error(GraphqlError::new(err.to_string(), ErrorCode::RateLimitError))
                    .first_error_message()
                    .unwrap_or("Internal server error".into()),
            );
        }

        let request_context = match self.create_request_context(headers).await {
            Ok(context) => context,
            Err(response) => return Err(response.first_error_message().unwrap_or("Internal server error".into())),
        };

        Ok(Session {
            engine: Arc::clone(self),
            request_context: Arc::new(request_context),
        })
    }

    async fn create_request_context(
        &self,
        headers: http::HeaderMap,
    ) -> Result<RequestContext<<R::Hooks as Hooks>::Context>, Response> {
        let client = Client::extract_from(&headers);
        let streaming_format = headers.typed_get::<StreamingFormat>();

        let (hooks_context, headers) = self
            .runtime
            .hooks()
            .on_gateway_request(headers)
            .await
            .map_err(Response::pre_execution_error)?;

        if let Some(access_token) = self.auth.authenticate(&headers).await {
            Ok(RequestContext {
                headers,
                streaming_format,
                client,
                access_token,
                hooks_context,
            })
        } else {
            Err(Response::pre_execution_error(GraphqlError::new(
                "Unauthenticated",
                ErrorCode::Unauthenticated,
            )))
        }
    }

    async fn execute_with_context(
        self: &Arc<Self>,
        request_context: RequestContext<<R::Hooks as Hooks>::Context>,
        batch_request: BatchRequest,
    ) -> HttpGraphqlResponse {
        match batch_request {
            BatchRequest::Single(request) => {
                if let Some(streaming_format) = request_context.streaming_format {
                    convert_stream_to_http_response(
                        streaming_format,
                        self.execute_stream(Arc::new(request_context), request),
                    )
                    .await
                } else {
                    self.execute_single(&request_context, request).await
                }
            }
            BatchRequest::Batch(requests) => {
                if request_context.streaming_format.is_some() {
                    return HttpGraphqlResponse::bad_request_error(
                        "batch requests can't use multipart or event-stream responses",
                    );
                }
                HttpGraphqlResponse::from_batch(
                    futures_util::stream::iter(requests.into_iter())
                        .then(|request| self.execute_single(&request_context, request))
                        .collect::<Vec<_>>()
                        .await,
                )
            }
        }
    }

    async fn execute_single(
        &self,
        request_context: &RequestContext<<R::Hooks as Hooks>::Context>,
        request: Request,
    ) -> HttpGraphqlResponse {
        let start = Instant::now();
        let span = GqlRequestSpan::create();
        async {
            let ctx = PreExecutionContext::new(self, request_context);
            let (operation_metadata, response) = ctx.execute_single(request).await;
            let status = response.status();
            let mut response_metadata = HttpGraphqlResponseExtraMetadata {
                operation_name: None,
                operation_type: None,
                has_errors: !status.is_success(),
            };
            if let Some(OperationMetadata {
                ty,
                name,
                normalized_query,
                normalized_query_hash,
            }) = operation_metadata
            {
                span.record_gql_request(GqlRequestAttributes {
                    operation_type: ty.as_str(),
                    operation_name: name.as_deref(),
                    sanitized_query: Some(&normalized_query),
                });
                response_metadata.operation_name.clone_from(&name);
                response_metadata.operation_type = Some(ty.as_str());
                self.operation_metrics.record(
                    GraphqlOperationMetricsAttributes {
                        ty: ty.as_str(),
                        name,
                        normalized_query,
                        normalized_query_hash,
                        status,
                        cache_status: None,
                        client: request_context.client.clone(),
                    },
                    start.elapsed(),
                );
            }
            span.record_gql_status(status);
            HttpGraphqlResponse::build(response, None, response_metadata)
        }
        .instrument(span.clone())
        .await
    }

    fn execute_stream(
        self: &Arc<Self>,
        request_context: Arc<RequestContext<<R::Hooks as Hooks>::Context>>,
        request: Request,
    ) -> impl Stream<Item = Response> + Send + 'static {
        let start = Instant::now();
        let engine = Arc::clone(self);
        let (sender, receiver) = mpsc::channel(2);

        let span = GqlRequestSpan::create();
        let span_clone = span.clone();
        receiver.join(
            async move {
                let ctx = PreExecutionContext::new(&engine, &request_context);
                let (metadata, status) = ctx.execute_stream(request, sender).await;
                if let Some(OperationMetadata {
                    ty,
                    name,
                    normalized_query,
                    normalized_query_hash,
                }) = metadata
                {
                    span.record_gql_request(GqlRequestAttributes {
                        operation_type: ty.as_str(),
                        operation_name: name.as_deref(),
                        sanitized_query: Some(&normalized_query),
                    });
                    engine.operation_metrics.record(
                        GraphqlOperationMetricsAttributes {
                            ty: ty.as_str(),
                            name,
                            normalized_query,
                            normalized_query_hash,
                            status,
                            cache_status: None,
                            client: request_context.client.clone(),
                        },
                        start.elapsed(),
                    );
                }

                span.record_gql_status(status);
            }
            .instrument(span_clone),
        )
    }
}

async fn convert_stream_to_http_response(
    streaming_format: StreamingFormat,
    stream: impl Stream<Item = Response> + Send + 'static,
) -> HttpGraphqlResponse {
    let mut stream = Box::pin(stream);
    let Some(first_response) = stream.next().await else {
        return HttpGraphqlResponse::internal_server_error("Empty stream");
    };
    HttpGraphqlResponse::from_stream(
        streaming_format,
        // Not perfect for the errors count, but good enough to detect a request error
        first_response.status(),
        futures_util::stream::iter(std::iter::once(first_response)).chain(stream),
    )
}

impl<'ctx, R: Runtime> PreExecutionContext<'ctx, R> {
    async fn execute_single(mut self, request: Request) -> (Option<OperationMetadata>, Response) {
        let operation_plan = match self.prepare_operation(request).await {
            Ok(operation_plan) => operation_plan,
            Err((metadata, response)) => return (metadata, response),
        };

        let metadata = Some(operation_plan.metadata.clone());
        let response = if matches!(operation_plan.ty(), OperationType::Subscription) {
            Response::pre_execution_error(GraphqlError::new(
                "Subscriptions are only suported on streaming transports. Try making a request with SSE or WebSockets",
                ErrorCode::BadRequest,
            ))
        } else {
            self.execute_query_or_mutation(operation_plan).await
        };

        (metadata, response)
    }

    async fn execute_stream(
        mut self,
        request: Request,
        mut sender: mpsc::Sender<Response>,
    ) -> (Option<OperationMetadata>, GraphqlResponseStatus) {
        let operation_plan = match self.prepare_operation(request).await {
            Ok(operation_plan) => operation_plan,
            Err((metadata, response)) => {
                let status = response.status();
                sender.send(response).await.ok();
                return (metadata, status);
            }
        };
        let operation_type = operation_plan.ty();
        let metadata = Some(operation_plan.metadata.clone());

        if matches!(operation_type, OperationType::Query | OperationType::Mutation) {
            let response = self.execute_query_or_mutation(operation_plan).await;
            let status = response.status();
            sender.send(response).await.ok();
            return (metadata, status);
        }

        let mut status: GraphqlResponseStatus = GraphqlResponseStatus::Success;
        struct Sender<'a> {
            sender: mpsc::Sender<Response>,
            status: &'a mut GraphqlResponseStatus,
        }

        impl crate::execution::ResponseSender for Sender<'_> {
            type Error = mpsc::SendError;
            async fn send(&mut self, response: Response) -> Result<(), Self::Error> {
                *self.status = self.status.union(response.status());
                self.sender.send(response).await
            }
        }

        self.execute_subscription(
            operation_plan,
            Sender {
                sender,
                status: &mut status,
            },
        )
        .await;
        (metadata, status)
    }

    async fn prepare_operation(
        &mut self,
        mut request: Request,
    ) -> Result<ExecutableOperation, (Option<OperationMetadata>, Response)> {
        let result = {
            let PreparedOperationDocument {
                cache_key,
                document_fut,
            } = match self.prepare_operation_document(&request) {
                Ok(pq) => pq,
                Err(err) => return Err((None, Response::pre_execution_error(err))),
            };
            if let Some(operation) = self.operation_cache.get(&cache_key).await {
                Ok(operation)
            } else if let Some(persisted_query) = document_fut {
                match persisted_query.await {
                    Ok(query) => Err((cache_key, Some(query))),
                    Err(err) => return Err((None, Response::pre_execution_error(err))),
                }
            } else {
                Err((cache_key, None))
            }
        };

        let operation = match result {
            Ok(operation) => operation,
            Err((cache_key, query)) => {
                if let Some(query) = query {
                    request.query = query
                }
                let operation = Operation::build(&self.schema, &request)
                    .map(Arc::new)
                    .map_err(|mut err| (err.take_operation_metadata(), Response::pre_execution_error(err)))?;

                self.push_background_future(self.engine.operation_cache.insert(cache_key, operation.clone()).boxed());
                operation
            }
        };

        let variables = Variables::build(self.schema.as_ref(), &operation, request.variables)
            .map_err(|errors| (Some(operation.metadata.clone()), Response::pre_execution_errors(errors)))?;

        self.finalize_operation(Arc::clone(&operation), variables)
            .await
            .map_err(|err| (Some(operation.metadata.clone()), Response::pre_execution_error(err)))
    }
}

pub struct Session<R: Runtime> {
    engine: Arc<Engine<R>>,
    request_context: Arc<RequestContext<<R::Hooks as Hooks>::Context>>,
}

impl<R: Runtime> Clone for Session<R> {
    fn clone(&self) -> Self {
        Self {
            engine: Arc::clone(&self.engine),
            request_context: Arc::clone(&self.request_context),
        }
    }
}

pub(crate) struct RequestContext<C> {
    pub headers: http::HeaderMap,
    pub streaming_format: Option<StreamingFormat>,
    pub client: Option<Client>,
    pub access_token: AccessToken,
    pub hooks_context: C,
}

impl<R: Runtime> Session<R> {
    pub fn execute_websocket(&self, id: String, request: Request) -> impl Stream<Item = websocket::Message> {
        self.engine
            .execute_stream(self.request_context.clone(), request)
            .map(move |response| match response {
                Response::PreExecutionError(_) => websocket::Message::Error {
                    id: id.clone(),
                    payload: websocket::Payload(response),
                },
                response => websocket::Message::Next {
                    id: id.clone(),
                    payload: websocket::Payload(response),
                },
            })
    }
}
