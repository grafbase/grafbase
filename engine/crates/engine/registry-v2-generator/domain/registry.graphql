# type SchemaDefinition @file(name: "schemas") {
#   description: StringLiteral
#   directives: [Directive!]!
#   root_operations: [RootOperationTypeDefinition!]!
# }

# type RootOperationTypeDefinition @file(name: "schemas") {
#   operation_type: OperationType!
#   named_type: String!
# }

# type ScalarDefinition @file(name: "scalars") {
#   name: String!
#   description: StringLiteral
#   directives: [Directive]
#   span: Span!
# }

union MetaType
  @distinct
  @file(name: "metatype")
  @variant(names: ["Object", "Interface", "Union", "Enum", "InputObject", "Scalar"]) =
    ObjectType
  | InterfaceType
  | UnionType
  | EnumType
  | InputObjectType
  | ScalarType

type ObjectType @file(name: "objects") @distinct {
  name: String!
  description: String
  fields: [MetaField!]!
  cache_control: CacheControl
  external: bool!
  shareable: bool!
  # directives: [Directive]
  # implements_interfaces: [String]
  # span: Span!
}

type MetaField @file(name: "field") {
  name: String!
  mapped_name: String
  description: String
  args: [MetaInputValue!]!
  ty: MetaFieldTypeRecord! @noreader
  deprecation: Deprecation
  cache_control: CacheControl
  requires: FieldSet
  federation: FederationProperties # inline Box this maybe?
  resolver: Resolver!
  required_operation: Operations
  auth: AuthConfig # inline box this also?
}

# TODO: Think about whether these should be inline or not...
# TODO: implement @box
scalar Deprecation @box @inline
scalar CacheControl @box @inline
scalar FieldSet @box @inline
scalar FederationProperties @box @inline
scalar Resolver @inline @ref
scalar Operations @box @inline
scalar AuthConfig @box @inline
scalar MetaFieldTypeRecord @inline

scalar ConstValue @inline @box

type MetaInputValue @file(name: "inputs") {
  name: String!
  description: String
  ty: MetaInputValueTypeRecord! @noreader
  default_value: ConstValue
  rename: String
  validators: [InputValidator!]!
}

scalar MetaInputValueTypeRecord @inline

type ScalarType @file(name: "scalar") @distinct {
  name: String!
  description: String
  specified_by_url: String
  parser: ScalarParser!
}

scalar ScalarParser @inline

type InterfaceType @file(name: "interface") @distinct {
  name: String!
  description: String
  fields: [MetaField!]!
  cache_control: CacheControl
  possible_types: [MetaType] @inline
}

type UnionType @file(name: "union") @distinct {
  name: String!
  description: String
  possible_types: [MetaType] @inline
  discriminators: UnionDiscriminators!
}

scalar UnionDiscriminators @inline @ref

type EnumType @file(name: "enums") @distinct {
  name: String!
  description: String
  values: [MetaEnumValue!]!
}

type MetaEnumValue @file(name: "enums") {
  name: String!
  description: String
  deprecation: Deprecation
  value: String
}

type InputObjectType @file(name: "inputs") @distinct {
  name: String!
  description: String
  input_fields: [MetaInputValue!]!
  oneof: bool!
}

type MetaDirective @file(name: "directives") @distinct {
  name: String!
  description: String
  locations: [DirectiveLocation!]!
  args: [MetaInputValue!]!
  is_repeatable: bool!
}

scalar DirectiveLocation @inline

type InputValidator @file(name: "inputs") {
  validator: DynValidator!
}

scalar DynValidator @inline @ref

# type FieldDefinition @file(name: "fields") {
#   name: String!
#   ty: Type!
#   arguments: [InputValueDefinition!]!
#   description: StringLiteral
#   directives: [Directive!]!
#   span: Span!
# }

# type InterfaceDefinition @file(name: "interfaces") {
#   name: String!
#   description: StringLiteral
#   fields: [FieldDefinition!]!
#   directives: [Directive!]!
#   implements_interfaces: [String]
#   span: Span!
# }

# type UnionDefinition @file(name: "unions") {
#   name: String!
#   description: StringLiteral
#   members: [String]
#   directives: [Directive!]!
#   span: Span!
# }

# type EnumDefinition @file(name: "enums") {
#   name: String!
#   description: StringLiteral
#   values: [EnumValueDefinition!]!
#   directives: [Directive!]!
#   span: Span!
# }

# type EnumValueDefinition @file(name: "enums") {
#   value: String!
#   description: StringLiteral
#   directives: [Directive!]!
#   span: Span!
# }

# type InputObjectDefinition @file(name: "input_objects") {
#   name: String!
#   description: StringLiteral
#   fields: [InputValueDefinition!]!
#   directives: [Directive!]!
#   span: Span!
# }

# type InputValueDefinition @file(name: "input_values") {
#   name: String!
#   ty: Type!
#   description: StringLiteral
#   default_value: Value
#   directives: [Directive!]!
#   span: Span!
# }

# type DirectiveDefinition @file(name: "directives") {
#   name: String!
#   description: StringLiteral
#   arguments: [InputValueDefinition!]!
#   is_repeatable: bool!
#   locations: [DirectiveLocation!]!
#   span: Span!
# }

# type Directive @file(name: "directives") {
#   name: String!
#   arguments: [Argument!]!
# }

# type Argument @file(name: "arguments") {
#   name: String!
#   value: Value!
# }

# scalar Span @inline
# scalar DirectiveLocation @inline
# scalar OperationType @inline

# Type & Value are kind of special cases that aren't worth automating
# so we make them scalars and implement them by hand
# scalar Type @file(name: "types")
# scalar Value @file(name: "value")

# String is built in, but easier to implement stuff if its just in the .graphql file
# It is also special cased a bit in the rust code
scalar String
scalar bool @inline

scalar WrapperTypes
