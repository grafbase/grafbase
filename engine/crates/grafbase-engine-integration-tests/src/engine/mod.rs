mod builder;

use std::{collections::HashMap, future::IntoFuture, sync::Arc};

use futures::{future::BoxFuture, Stream, StreamExt};
use grafbase_engine::{Request, RequestHeaders, Response, Schema, StreamingPayload, Variables};

pub use self::builder::EngineBuilder;

/// An instance of the grafbase-engine code that can be used for testing.
#[derive(Clone)]
pub struct Engine {
    inner: Arc<Inner>,
}

struct Inner {
    schema: Schema,
}

impl Engine {
    pub async fn new(schema: String) -> Self {
        EngineBuilder::new(schema).build().await
    }

    pub fn execute(&self, operation: impl Into<GraphQlRequest>) -> ExecutionRequest<'_> {
        ExecutionRequest {
            graphql: operation.into(),
            headers: HashMap::new(),
            schema: &self.inner.schema,
        }
    }

    pub fn execute_stream(&self, operation: impl Into<GraphQlRequest>) -> StreamExecutionRequest<'_> {
        StreamExecutionRequest {
            graphql: operation.into(),
            headers: HashMap::new(),
            schema: &self.inner.schema,
        }
    }
}

#[must_use]
pub struct ExecutionRequest<'a> {
    graphql: GraphQlRequest,
    headers: HashMap<String, String>,
    schema: &'a Schema,
}

impl ExecutionRequest<'_> {
    /// Adds a header into the request
    pub fn header(mut self, name: impl Into<String>, value: impl Into<String>) -> Self {
        self.headers.insert(name.into(), value.into());
        self
    }
}

impl<'a> IntoFuture for ExecutionRequest<'a> {
    type Output = Response;

    type IntoFuture = BoxFuture<'a, Response>;

    fn into_future(self) -> Self::IntoFuture {
        Box::pin(async move {
            let mut request = Request::new(self.graphql.query).data(RequestHeaders::from(&self.headers));
            if let Some(name) = self.graphql.operation_name {
                request = request.operation_name(name);
            }
            if let Some(variables) = self.graphql.variables {
                request = request.variables(variables);
            }
            self.schema.execute(request).await
        })
    }
}

#[must_use]
pub struct StreamExecutionRequest<'a> {
    graphql: GraphQlRequest,
    headers: HashMap<String, String>,
    schema: &'a Schema,
}

impl<'a> StreamExecutionRequest<'a> {
    /// Adds a header into the request
    pub fn header(mut self, name: impl Into<String>, value: impl Into<String>) -> Self {
        self.headers.insert(name.into(), value.into());
        self
    }

    /// Converts the execution request into a Stream
    pub fn into_stream(self) -> impl Stream<Item = StreamingPayload> + 'a {
        let mut request = Request::new(self.graphql.query).data(RequestHeaders::from(&self.headers));
        if let Some(name) = self.graphql.operation_name {
            request = request.operation_name(name);
        }
        if let Some(variables) = self.graphql.variables {
            request = request.variables(variables);
        }
        self.schema.execute_stream(request)
    }

    // Collects the StreamingPayloads into a vec
    pub async fn collect(self) -> Vec<StreamingPayload> {
        self.into_stream().collect().await
    }

    /// Converts the execution request into an iterator
    pub async fn into_iter(self) -> impl Iterator<Item = StreamingPayload> {
        self.collect().await.into_iter()
    }
}

pub struct GraphQlRequest {
    query: String,
    operation_name: Option<String>,
    variables: Option<Variables>,
}

impl From<&str> for GraphQlRequest {
    fn from(val: &str) -> Self {
        GraphQlRequest {
            query: val.into(),
            operation_name: None,
            variables: None,
        }
    }
}

impl From<String> for GraphQlRequest {
    fn from(val: String) -> Self {
        GraphQlRequest {
            query: val,
            operation_name: None,
            variables: None,
        }
    }
}

impl<T, V> From<cynic::Operation<T, V>> for GraphQlRequest
where
    V: serde::Serialize,
{
    fn from(operation: cynic::Operation<T, V>) -> Self {
        GraphQlRequest {
            query: operation.query,
            variables: Some(serde_json::from_value(serde_json::to_value(operation.variables).unwrap()).unwrap()),
            operation_name: operation.operation_name.map(|name| name.to_string()),
        }
    }
}
