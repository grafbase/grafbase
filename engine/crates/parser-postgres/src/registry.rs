mod context;
mod queries;
mod types;

use context::InputContext;
use parser_sdl::Registry;
use postgres_connector_types::database_definition::DatabaseDefinition;

use self::context::OutputContext;

pub(super) fn generate(database_definition: DatabaseDefinition, name: &str, namespaced: bool) -> Registry {
    let input_ctx = InputContext::new(database_definition, name, namespaced);
    let mut output_ctx = OutputContext::new(namespaced.then_some(name));

    types::generate(&input_ctx, &mut output_ctx);
    queries::generate(&input_ctx, &mut output_ctx);

    output_ctx.finalize(input_ctx.finalize(), name)
}
