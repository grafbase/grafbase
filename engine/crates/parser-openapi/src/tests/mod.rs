use std::sync::Once;

use assert_matches::assert_matches;
use engine::registry::{MetaType, UnionType};

use super::*;

mod federation;

#[test]
fn test_stripe_output() {
    let metadata = ApiMetadata {
        url: None,
        ..metadata("stripe", true)
    };
    insta::assert_snapshot!(build_registry("test_data/stripe.openapi.json", Format::Json, metadata)
        .unwrap()
        .export_sdl(false));
}

#[test]
fn test_stripe_output_json() {
    let metadata = ApiMetadata {
        url: None,
        ..metadata("stripe", true)
    };
    let registry = build_registry("test_data/stripe.openapi.json", Format::Json, metadata).unwrap();
    insta::assert_json_snapshot!(registry);
}

#[test]
fn test_petstore_output() {
    let registry = build_registry(
        "test_data/petstore.openapi.json",
        Format::Json,
        metadata("petstore", true),
    )
    .unwrap();

    insta::assert_snapshot!(registry.export_sdl(false));
    insta::assert_debug_snapshot!(registry);
}

#[test]
fn test_petstore_without_prefix() {
    let registry = build_registry(
        "test_data/petstore.openapi.json",
        Format::Json,
        ApiMetadata {
            type_prefix: None,
            ..metadata("petstore", false)
        },
    )
    .unwrap();

    insta::assert_snapshot!(registry.export_sdl(false));
}

#[test]
fn test_flat_output() {
    let registry = build_registry(
        "test_data/petstore.openapi.json",
        Format::Json,
        metadata("petstore", false),
    )
    .unwrap();

    insta::assert_snapshot!(registry.export_sdl(false));
    insta::assert_debug_snapshot!(registry);
}

#[test]
fn test_url_without_host_failure() {
    let metadata = ApiMetadata {
        url: None,
        ..metadata("petstore", true)
    };
    assert_matches!(
        build_registry("test_data/petstore.openapi.json", Format::Json, metadata)
            .unwrap_err()
            .as_slice(),
        [Error::InvalidUrl(url)] => {
            assert_eq!(url, "/api/v3");
        }
    );
}

#[test]
fn test_openai_output() {
    insta::assert_snapshot!(build_registry(
        "test_data/openai.yaml",
        Format::Yaml,
        ApiMetadata {
            query_naming: QueryNamingStrategy::OperationId,
            ..metadata("openai", true)
        }
    )
    .unwrap()
    .export_sdl(false));
}

#[test]
fn test_planetscale() {
    insta::assert_snapshot!(build_registry(
        "test_data/planetscale.json",
        Format::Json,
        ApiMetadata {
            url: None,
            ..metadata("planetscale", true)
        }
    )
    .unwrap()
    .export_sdl(false));
}

#[test]
fn test_orb() {
    // Orb is a 3.1 spec
    insta::assert_snapshot!(build_registry(
        "test_data/orb.json",
        Format::Json,
        ApiMetadata {
            url: None,
            ..metadata("orb", true)
        }
    )
    .unwrap()
    .export_sdl(false));
}

#[test]
fn test_mongo_atlas() {
    init_tracing();

    // Mongo Atlas is a 3.1 spec
    insta::assert_snapshot!(build_registry(
        "test_data/mongo-atlas.json",
        Format::Json,
        ApiMetadata {
            url: None,
            ..metadata("mongo", true)
        }
    )
    .unwrap()
    .export_sdl(false));
}

#[test]
fn test_impossible_unions() {
    insta::assert_snapshot!(build_registry(
        "test_data/impossible-unions.json",
        Format::Json,
        metadata("petstore", true)
    )
    .unwrap()
    .export_sdl(false));
}

#[test]
fn test_all_of_schema_simple() {
    // an allOf schema that are simple merges of distinct objects
    insta::assert_snapshot!(build_registry(
        "test_data/all-ofs-simple.json",
        Format::Json,
        metadata("petstore", true)
    )
    .unwrap()
    .export_sdl(false));
}

#[test]
fn test_all_of_schema_complex() {
    // Some allOf schemas have properties defined in multiple branches
    // of the allOf, with required sometiems being in one branch but not
    // another.  This is a test of that....
    insta::assert_snapshot!(build_registry(
        "test_data/all-ofs-complex.json",
        Format::Json,
        metadata("petstore", true)
    )
    .unwrap()
    .export_sdl(false));
}

#[test]
fn test_supabase() {
    insta::assert_snapshot!(
        build_registry("test_data/supabase.json", Format::Json, metadata("supabase", true))
            .unwrap()
            .export_sdl(false)
    );
}

#[test]
fn test_greenlake_schema() {
    init_tracing();

    let metadata = ApiMetadata {
        url: None,
        query_naming: QueryNamingStrategy::OperationId,
        ..metadata("greenlake", false)
    };

    let registry = build_registry("test_data/greenlake.yml", Format::Json, metadata).unwrap();

    insta::assert_snapshot!(registry.export_sdl(false));
}

#[test]
fn test_stripe_discrimnator_detection() {
    let registry = build_registry("test_data/stripe.openapi.json", Format::Json, metadata("stripe", true)).unwrap();
    let discriminators = registry
        .types
        .values()
        .filter_map(|ty| match ty {
            MetaType::Union(UnionType {
                name, discriminators, ..
            }) => Some((name, discriminators)),
            _ => None,
        })
        .collect::<Vec<_>>();

    insta::assert_json_snapshot!(discriminators);
}

fn build_registry(schema_path: &str, format: Format, metadata: ApiMetadata) -> Result<Registry, Vec<Error>> {
    let mut registry = Registry::new();

    parse_spec(
        std::fs::read_to_string(schema_path).unwrap(),
        format,
        metadata,
        &mut registry,
    )?;

    Ok(registry)
}

fn metadata(name: &str, namespace: bool) -> ApiMetadata {
    ApiMetadata {
        name: name.to_string(),
        namespace,
        url: Some(Url::parse("http://example.com").unwrap()),
        headers: ConnectorHeaders::new([]),
        query_naming: QueryNamingStrategy::SchemaName,
        type_prefix: Some(name.to_string()),
    }
}

fn init_tracing() {
    static INITIALIZER: Once = Once::new();
    INITIALIZER.call_once(|| {
        tracing_subscriber::fmt()
            .with_env_filter("trace")
            .without_time()
            .with_test_writer()
            .init();
    });
}
