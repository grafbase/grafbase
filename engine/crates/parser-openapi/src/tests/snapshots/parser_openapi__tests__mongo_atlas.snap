---
source: engine/crates/parser-openapi/src/tests/mod.rs
expression: "build_registry(\"test_data/mongo-atlas.json\", Format::Json,\n            ApiMetadata {\n                url: None,\n                ..metadata(\"mongo\", true)\n            }).unwrap().export_sdl(false)"
---
type MongoAggregate {
	documents: [JSON!]!
}
input MongoAggregateInput {
	pipeline: [JSON!]
	database: String!
	dataSource: String!
	collection: String!
}
type MongoDeleteMany {
	deletedCount: Float!
}
input MongoDeleteManyInput {
	filter: JSON
	database: String!
	dataSource: String!
	collection: String!
}
type MongoDeleteOne {
	deletedCount: Float!
}
input MongoDeleteOneInput {
	filter: JSON
	database: String!
	dataSource: String!
	collection: String!
}
input MongoFindInput {
	sort: JSON
	skip: Float
	projection: JSON
	limit: Float
	filter: JSON
	database: String!
	dataSource: String!
	collection: String!
}
type MongoFindManyResponseBody {
	documents: [JSON!]
}
input MongoFindOneInput {
	projection: JSON
	filter: JSON
	database: String!
	dataSource: String!
	collection: String!
}
type MongoFindOneResponseBody {
	document: JSON
}
type MongoInsertMany {
	insertedIds: [String!]
}
input MongoInsertManyInput {
	documents: [JSON!]!
	database: String!
	dataSource: String!
	collection: String!
}
type MongoInsertOne {
	insertedId: String
}
input MongoInsertOneInput {
	document: JSON!
	database: String!
	dataSource: String!
	collection: String!
}
type MongoMutation {
	findOne(input: MongoFindOneInput): MongoFindOneResponseBody
	find(input: MongoFindInput): MongoFindManyResponseBody
	insertOne(input: MongoInsertOneInput): MongoInsertOne
	insertMany(input: MongoInsertManyInput): MongoInsertMany
	updateOne(input: MongoUpdateOneInput): MongoUpdateOne
	updateMany(input: MongoUpdateManyInput): MongoUpdateMany
	deleteOne(input: MongoDeleteOneInput): MongoDeleteOne
	deleteMany(input: MongoDeleteManyInput): MongoDeleteMany
	aggregate(input: MongoAggregateInput): MongoAggregate
}
type MongoUpdateMany {
	upsertedId: String
	modifiedCount: Float!
	matchedCount: Float!
}
input MongoUpdateManyInput {
	upsert: Boolean = false
	update: JSON!
	filter: JSON
	database: String!
	dataSource: String!
	collection: String!
}
type MongoUpdateOne {
	upsertedId: String
	modifiedCount: Float!
	matchedCount: Float!
}
input MongoUpdateOneInput {
	upsert: Boolean = false
	update: JSON!
	filter: JSON
	database: String!
	dataSource: String!
	collection: String!
}
type Mutation {
	mongo: MongoMutation!
}
type Query {
}
schema {
	query: Query
	mutation: Mutation
}
